File: src/activation.rs
pub struct Activation {}

impl Activation {
    pub fn sigmoid(x: f64) -> f64 {
        1.0 / (1.0 + (-x).exp())
    }

    pub fn sigmoid_derivative(x: f64) -> f64 {
        x * (1.0 - x)
    }

    pub fn relu(x: f64) -> f64 {
        if x > 0.0 {
            return x;
        } else {
            return 0.0;
        }
    }

    pub fn relu_derivative(x: f64) -> f64 {
        if x > 0.0 {
            return 1.0;
        } else {
            return 0.0;
        }
    }

    pub fn tanh_derivative(x: f64) -> f64 {
        1.0 - (x * x)
    }
}

File: src/math/activation.cpp
#include "math/activation.h"

double sigmoid(double x) {
     return 1/(1-exp(-x));
}

double d_sigmoid(double x) {
    return sigmoid(x)*(1-sigmoid(x));
}

double relu(double x) {
    return x > 0 ? x : 0;
}

double d_relu(double x) {
    return x > 0 ? 1 : 0;
}

double linear(double x) {
    return x;
}

double d_linear(double x) {
    return 1;
}

double tanh(double x) {
    return std::tanh(x);
}

double d_tanh(double x) {
    return 1 - std::pow(tanh(x), 2);
}